core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1505,6 +1505,72 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     }
   }
 
+  def groupWithin[F2[x] >: F[x]](
+      chunkSize: Int,
+      timeout: FiniteDuration
+  )(implicit F: Temporal[F2]): Stream[F2, Chunk[O]] = {
+
+    val groupSize: Long = chunkSize.toLong
+
+    if (timeout.toNanos == 0) chunkN(chunkSize)
+    else
+      Stream.force {
+        for {
+          supply <- Semaphore[F2](0)
+          supplyEnded <- Ref.of(false)
+          buffer <- Queue.bounded[F2, O](chunkSize) // buffering and backpressure
+        } yield {
+
+          val emitChunk: F2[Chunk[O]] =
+            buffer.tryTakeN(Some(groupSize.toInt)).map(Chunk.seq)
+
+          // we need to check the buffer size, rather than the available supply since
+          // the supply is maxed out at the end so won't always indicate accurately the buffer size
+          val isBufferEmpty: F2[Boolean] =
+            buffer.size.map(_ == 0)
+
+          val streamExhausted: F2[Boolean] =
+            (isBufferEmpty, supplyEnded.get).mapN(_ && _)
+
+          //  releasing a number of permits equal to {groupSize} should be enough,
+          //  but in order to ensure termination of the consumer when the producer
+          //  stream is cancelled we need to max out the supply
+          val maxOutSupply: F2[Unit] =
+            supply.available.flatMap(a => supply.releaseN(Long.MaxValue - a))
+
+          // enabling termination of the consumer stream when the producer completes
+          // naturally (i.e runs out of elements)
+          val markSupplyEnd: F2[Unit] = supplyEnded.set(true)
+
+          val enqueue: F2[Unit] =
+            foreach(buffer.offer(_) <* supply.release).compile.drain
+              .guarantee(markSupplyEnd *> maxOutSupply)
+
+          val awaitAndEmitNext: F2[Chunk[O]] = for {
+            isEmpty <- isBufferEmpty
+            awaited <- supply.acquire.whenA(isEmpty).as(if (isEmpty) 1 else 0)
+            flushed <- emitChunk
+            // lower supply by {flushed.size} (excluding element already awaited)
+            _ <- supply.tryAcquireN((flushed.size.toLong - awaited).max(0))
+          } yield flushed
+
+          val onTimeout =
+            F.ifM(streamExhausted)(F.pure(Chunk.empty[O]), awaitAndEmitNext)
+
+          val dequeue: F2[Chunk[O]] =
+            F.race(supply.acquireN(groupSize), F.sleep(timeout)).flatMap {
```

Comment:
> I’ve read the issue attached above (race condition), am I right to think that this implementation is potentially better if the race is left-biased, since the timeout logic (on the right) is slightly more convoluted ?

Less chance of deadlock is still not as good as no chance of deadlock 😁 

I think the currently implementation uses a hack, where it always returns the semaphore, and then attempts to re-acquire it. But there's no way to fix this properly without implementing the `Ior`-based `race` as described in that issue.

Reply:
I was struggling a bit to understand the problem (i.e where the deadlock would occur) until I clicked through a few links from the issue linked above and stumbled across your PR comment [here](https://github.com/typelevel/cats-effect/pull/3233#issue-1439117768)

If I understood your concern correctly the issue is that the semaphore might lose the race, even if it acquired the permits successfully. So in that case we would need to restore them, in order to prevent a deadlock when the `onTimeout` logic tries to acquire a permit again (that simply won't be there if the semaphore acquired them without releasing them)

if that is the case I believe this is not a problem in this implementation because of this logic:

```scala
     val isBufferEmpty: F2[Boolean] =
        buffer.size.map(_ == 0)

      val awaitAndEmitNext: F2[Chunk[O]] = for {
        isEmpty <- isBufferEmpty
        awaited <- supply.acquire.whenA(isEmpty).as(if (isEmpty) 1 else 0)
        flushed <- emitChunk
        // lower supply by {flushed.size} (excluding element already awaited)
        _ <- supply.tryAcquireN((flushed.size.toLong - awaited).max(0))
      } yield flushed
``` 

basically if the buffer contains any element by the time the timeout expires we won't attempt to acquire the permit in a blocking fashion in the for-comprehension. 
So by definition the deadlock cannot occur or in other words we should be able to rule out the deadlock because if the buffer is empty then we will know for a fact that the competing semaphore (the one in the race) hasn't acquired any permit (i.e. it does not need to release them)

Reply:
Thanks for looking more closely at this! Yes, you've definitely understood the issue.

After looking at your code and thinking about it, I _think_ you are right that it won't deadlock. However, I still suspect that the state can get corrupted. Something like this:

1. The `race` times-out, although the semaphore `acquireN`  also succeeds.
https://github.com/typelevel/fs2/blob/d676eddbce407908b7a6daaa00666b88b968cd0c/core/shared/src/main/scala/fs2/Stream.scala#L1456-L1458
2. We successfully `emitChunk` via `onTimeout` and `awaitAndEmitNext`.
https://github.com/typelevel/fs2/blob/d676eddbce407908b7a6daaa00666b88b968cd0c/core/shared/src/main/scala/fs2/Stream.scala#L1452-L1453
https://github.com/typelevel/fs2/blob/d676eddbce407908b7a6daaa00666b88b968cd0c/core/shared/src/main/scala/fs2/Stream.scala#L1444-L1447
https://github.com/typelevel/fs2/blob/d676eddbce407908b7a6daaa00666b88b968cd0c/core/shared/src/main/scala/fs2/Stream.scala#L1420-L1421
3. This unblocks the `buffer`, which concurrently gets filled with new stuff and releases permits.
https://github.com/typelevel/fs2/blob/d676eddbce407908b7a6daaa00666b88b968cd0c/core/shared/src/main/scala/fs2/Stream.scala#L1441-L1442
4. The `tryAcquireN` in `awaitAndEmitNext` succeeds.
https://github.com/typelevel/fs2/blob/d676eddbce407908b7a6daaa00666b88b968cd0c/core/shared/src/main/scala/fs2/Stream.scala#L1448-L1450
5. At this point, we've consumed `2*chunkSize` permits to emit `chunkSize` elements.

Reply:
Nice!! Using `TestControl.executeEmbed` is a great idea for this! Will give it a go 👍🏾